src/ray/raylet/local_object_manager.cc

```diff
@@ -40,6 +40,9 @@ void LocalObjectManager::PinObjects(const std::vector<ObjectID> &object_ids,
 void LocalObjectManager::WaitForObjectFree(const rpc::Address &owner_address,
                                            const std::vector<ObjectID> &object_ids) {
   for (const auto &object_id : object_ids) {
+    // Save the object's owner in case we need to communicate with the owner upon an
+    // object spill.
+    pinned_object_owners_.emplace(object_id, owner_address);
```

Comment:
Could we squash this hashmap with the existing `pinned_objects_`/`objects_pending_spill_` to avoid having to insert/delete from multiple hashmaps? That way we can avoid memory leaks.

Reply:
I was hoping to merge this hashmap into `pinned_objects_` and `objects_pending_spill_`, but I was concerned that the lifecycles may not line up. E.g., that we'd want the owner entries to live after spilling and only be deleted when the reference is freed. Am I right to be concerned there, or do you think that the lifecycles line up?

Reply:
Ah I see, that makes sense. Yes, I didn't realize at the time that we need to keep the owner address after the object is spilled.

Reply:
Hmm actually, I don't think that we should be keeping the owner address here, right? Because the object could also get restored on other nodes that don't have this entry in their local map. It seems like we'll need to pass the owner address elsewhere.

In that case, could you actually squash the hashmaps after all? I think it will be best to try and make sure that the lifecycles line up.

Reply:
Sure I can try to make them line up!